## <font color = Navy > __Churn Model__ </font>
#### Case :
Trying to predict whether a customer in a bank is more likely to leave the bank or stay with it using the past record of 10000 customers.

In [26]:
import pandas as pd
import keras
import numpy as np

In [2]:
data = pd.read_csv('./Churn_Modelling.csv', sep = ',')
len(data)

10000

In [3]:
data[:10]

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   
5          6    15574012       Chu          645     Spain    Male   44   
6          7    15592531  Bartlett          822    France    Male   50   
7          8    15656148    Obinna          376   Germany  Female   29   
8          9    15792365        He          501    France    Male   44   
9         10    15592389        H?          684    France    Male   27   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   
5       8  113755.78              2          1               0   
6       7       0.00              2          1               1   
7       4  115046.74              4          1               0   
8       4  142051.07              2          0               1   
9       2  134603.88              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0  
5        149756.71       1  
6         10062.80       0  
7        119346.88       1  
8         74940.50       0  
9         71725.73       0

### *Feature Matrix*

In [4]:
X = data.iloc[:,3:-1].values
X[:10]

array([[619, 'France', 'Female', 42, 2, 0.0, 1, 1, 1, 101348.88],
       [608, 'Spain', 'Female', 41, 1, 83807.86, 1, 0, 1, 112542.58],
       [502, 'France', 'Female', 42, 8, 159660.8, 3, 1, 0, 113931.57],
       [699, 'France', 'Female', 39, 1, 0.0, 2, 0, 0, 93826.63],
       [850, 'Spain', 'Female', 43, 2, 125510.82, 1, 1, 1, 79084.1],
       [645, 'Spain', 'Male', 44, 8, 113755.78, 2, 1, 0, 149756.71],
       [822, 'France', 'Male', 50, 7, 0.0, 2, 1, 1, 10062.8],
       [376, 'Germany', 'Female', 29, 4, 115046.74, 4, 1, 0, 119346.88],
       [501, 'France', 'Male', 44, 4, 142051.07, 2, 0, 1, 74940.5],
       [684, 'France', 'Male', 27, 2, 134603.88, 1, 1, 1, 71725.73]],
      dtype=object)

### *Labels*

In [5]:
Y = data.iloc[:,-1].values
Y[:10]


array([1, 0, 1, 0, 0, 1, 0, 1, 0, 0])

##### Preprocessing:
1. Encoding the Categorical data
2. Using standardisation on encoded dataset

In [13]:
from sklearn.preprocessing import StandardScaler,LabelEncoder,OneHotEncoder
scl = StandardScaler()
label_x_1 = LabelEncoder()
oneHot = OneHotEncoder(categorical_features=[1])
X[:,1] = label_x_1.fit_transform(X[:,1])
X[:,2] = label_x_1.fit_transform(X[:,2])

In [14]:
X = oneHot.fit_transform(X).toarray()

/home/pratik/.local/lib/python3.6/site-packages/sklearn/preprocessing/_encoders.py:368: FutureWarning: The handling of integer data will change in version 0.22. Currently, the categories are determined based on the range [0, max(values)], while in the future they will be determined based on the unique values.
If you want the future behaviour and silence this warning, you can specify "categories='auto'".
In case you used a LabelEncoder before this OneHotEncoder to convert the categories to integers, then you can now use the OneHotEncoder directly.
  warnings.warn(msg, FutureWarning)
/home/pratik/.local/lib/python3.6/site-packages/sklearn/preprocessing/_encoders.py:390: DeprecationWarning: The 'categorical_features' keyword is deprecated in version 0.20 and will be removed in 0.22. You can use the ColumnTransformer instead.
  "use the ColumnTransformer instead.", DeprecationWarning)


In [29]:
X[1]

array([ 1.74273971, -0.57873591, -0.44133106, -1.09598752,  0.19816383,
       -1.38753759,  0.11735002, -0.91158349, -1.54776799,  0.97024255,
        0.21653375])

In [16]:
X = X[:,1:]

In [17]:
X = scl.fit_transform(X)

In [18]:
X[:2]

array([[-0.57380915, -0.57873591, -0.32733113, -1.09598752,  0.29351742,
        -1.04175968, -1.22584767, -0.91158349,  0.64609167,  0.97024255,
         0.02188649],
       [ 1.74273971, -0.57873591, -0.44133106, -1.09598752,  0.19816383,
        -1.38753759,  0.11735002, -0.91158349, -1.54776799,  0.97024255,
         0.21653375]])

In [19]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(X, Y, train_size = 0.8, shuffle = True )

/home/pratik/.local/lib/python3.6/site-packages/sklearn/model_selection/_split.py:2179: FutureWarning: From version 0.21, test_size will always complement train_size unless both are specified.
  FutureWarning)


In [20]:
from keras.layers import Dense, Dropout
from keras.models import Sequential

In [21]:
classifier = Sequential()
classifier.add(Dense(input_dim = 11, units = 6, activation = 'relu', kernel_initializer = 'uniform' ))

classifier.add(Dropout(rate = 0.1))
classifier.add(Dense(units = 8, activation = 'relu', kernel_initializer = 'uniform' ))
classifier.add(Dropout(rate = 0.2))
# classifier.add(Dense(units = 4, activation = 'relu', kernel_initializer = 'uniform' ))
# classifier.add(Dropout(rate = 0.2))
classifier.add(Dense(units = 1, activation = 'sigmoid', kernel_initializer = 'uniform' ))

classifier.compile(optimizer = 'Adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

Instructions for updating:
Colocations handled automatically by placer.
Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.


In [22]:
classifier.fit(x_train, y_train, batch_size=50, epochs=150, validation_split=0.2, shuffle = True)

Instructions for updating:
Use tf.cast instead.
Train on 6400 samples, validate on 1600 samples
Epoch 1/150
6400/6400 [==============================] - 3s 475us/step - loss: 0.6420 - acc: 0.7944 - val_loss: 0.5323 - val_acc: 0.8094
Epoch 2/150
6400/6400 [==============================] - 1s 94us/step - loss: 0.4773 - acc: 0.7947 - val_loss: 0.4299 - val_acc: 0.8094
Epoch 3/150
6400/6400 [==============================] - 1s 91us/step - loss: 0.4523 - acc: 0.7947 - val_loss: 0.4236 - val_acc: 0.8094
Epoch 4/150
6400/6400 [==============================] - 1s 96us/step - loss: 0.4450 - acc: 0.7947 - val_loss: 0.4205 - val_acc: 0.8094
Epoch 5/150
6400/6400 [==============================] - 1s 105us/step - loss: 0.4432 - acc: 0.7947 - val_loss: 0.4176 - val_acc: 0.8094
Epoch 6/150
6400/6400 [==============================] - 1s 115us/step - loss: 0.4402 - acc: 0.7947 - val_loss: 0.4164 - val_acc: 0.8094
Epoch 7/150
6400/6400 [==============================] - 1s 93us/step - loss: 0.4400 

Epoch 118/150
6400/6400 [==============================] - 0s 71us/step - loss: 0.4269 - acc: 0.8305 - val_loss: 0.4005 - val_acc: 0.8487
Epoch 119/150
6400/6400 [==============================] - 0s 70us/step - loss: 0.4278 - acc: 0.8297 - val_loss: 0.4001 - val_acc: 0.8456
Epoch 120/150
6400/6400 [==============================] - 0s 75us/step - loss: 0.4269 - acc: 0.8297 - val_loss: 0.4011 - val_acc: 0.8487
Epoch 121/150
6400/6400 [==============================] - 0s 69us/step - loss: 0.4264 - acc: 0.8309 - val_loss: 0.4005 - val_acc: 0.8494
Epoch 122/150
6400/6400 [==============================] - 0s 68us/step - loss: 0.4271 - acc: 0.8302 - val_loss: 0.4003 - val_acc: 0.8475
Epoch 123/150
6400/6400 [==============================] - 0s 65us/step - loss: 0.4299 - acc: 0.8289 - val_loss: 0.4001 - val_acc: 0.8462
Epoch 124/150
6400/6400 [==============================] - 0s 65us/step - loss: 0.4275 - acc: 0.8300 - val_loss: 0.3997 - val_acc: 0.8475
Epoch 125/150
6400/6400 [=========

In [38]:
y_pred = classifier.predict(x_test)

In [39]:
y_pred = (y_pred > 0.5)

In [40]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

In [41]:
cm

array([[1543,   39],
       [ 303,  115]])

### <font color = Navy >_New Prediction_</font>

Predict if the customer with the following informations will leave the bank:  
Geography: France  
Credit Score: 800  
Gender: Female  
Age: 32  
Tenure: 2  
Balance: 120000  
Number of Products: 1  
Has Credit Card: Yes  
Is Active Member: Yes  
Estimated Salary: 140000  

In [44]:
new_pred = classifier.predict(scl.transform(np.array([[0, 0, 800, 0, 32, 2,120000, 1, 1, 1, 140000 ]])))
new_pred 

/home/pratik/.local/lib/python3.6/site-packages/sklearn/utils/validation.py:595: DataConversionWarning: Data with input dtype int64 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)


array([[0.]], dtype=float32)

Thus, the customer stayed with the bank.

### Improving the validity of results using cross-validation 

In [46]:
from keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import cross_val_score
def build_classifier():
    classifier = Sequential()
    # input Layer and first inner layer with 6 nodes
    classifier.add(Dense(input_dim = 11, units = 6, activation = 'relu', kernel_initializer = 'uniform' ))
    classifier.add(Dropout(rate = 0.1))
    # second Layer 
    classifier.add(Dense(units = 8, activation = 'relu', kernel_initializer = 'uniform' ))
    classifier.add(Dropout(rate = 0.2))
    # output Layer
    classifier.add(Dense(units = 1, activation = 'sigmoid', kernel_initializer = 'uniform' ))

    classifier.compile(optimizer = 'Adam', loss = 'binary_crossentropy', metrics = ['accuracy'])
    return classifier


In [30]:
classifier = KerasClassifier(build_fn=build_classifier,batch_size=50, epochs=150, shuffle = True )

In [87]:
accuracy = cross_val_score(estimator=classifier, X=x_train, y=y_train, cv=10, n_jobs=-1)

Epoch 1/150
7200/7200 [==============================] - 1s 134us/step - loss: 0.6432 - acc: 0.7932
Epoch 2/150
7200/7200 [==============================] - 0s 39us/step - loss: 0.4790 - acc: 0.7962
Epoch 3/150
7200/7200 [==============================] - 0s 39us/step - loss: 0.4436 - acc: 0.7967
Epoch 4/150
7200/7200 [==============================] - 0s 40us/step - loss: 0.4295 - acc: 0.8110
Epoch 5/150
7200/7200 [==============================] - 0s 39us/step - loss: 0.4264 - acc: 0.8217
Epoch 6/150
7200/7200 [==============================] - 0s 40us/step - loss: 0.4253 - acc: 0.8200
Epoch 7/150
7200/7200 [==============================] - 0s 39us/step - loss: 0.4233 - acc: 0.8207
Epoch 8/150
7200/7200 [==============================] - 0s 40us/step - loss: 0.4217 - acc: 0.8224
Epoch 9/150
7200/7200 [==============================] - 1s 85us/step - loss: 0.4130 - acc: 0.8237
Epoch 10/150
7200/7200 [==============================] - 1s 90us/step - loss: 0.4149 - acc: 0.8242
Epoch 11

Epoch 13/150
7200/7200 [==============================] - 1s 87us/step - loss: 0.4217 - acc: 0.8224
Epoch 14/150
7200/7200 [==============================] - 1s 79us/step - loss: 0.4236 - acc: 0.8197
Epoch 15/150
7200/7200 [==============================] - 1s 83us/step - loss: 0.4231 - acc: 0.8211
Epoch 16/150
7200/7200 [==============================] - 1s 72us/step - loss: 0.4201 - acc: 0.8214
Epoch 17/150
7200/7200 [==============================] - 1s 79us/step - loss: 0.4209 - acc: 0.8237
Epoch 18/150
7200/7200 [==============================] - 1s 78us/step - loss: 0.4193 - acc: 0.8222
Epoch 19/150
7200/7200 [==============================] - 0s 33us/step - loss: 0.4197 - acc: 0.8222
Epoch 20/150
7200/7200 [==============================] - 0s 54us/step - loss: 0.4219 - acc: 0.8210
Epoch 21/150
7200/7200 [==============================] - 0s 41us/step - loss: 0.4225 - acc: 0.8218
Epoch 22/150
7200/7200 [==============================] - 0s 35us/step - loss: 0.4176 - acc: 0.8192


Epoch 25/150
7200/7200 [==============================] - 0s 54us/step - loss: 0.4263 - acc: 0.8247
Epoch 26/150
7200/7200 [==============================] - 0s 37us/step - loss: 0.4285 - acc: 0.8271
Epoch 27/150
7200/7200 [==============================] - 0s 53us/step - loss: 0.4294 - acc: 0.8269
Epoch 28/150
7200/7200 [==============================] - 0s 37us/step - loss: 0.4279 - acc: 0.8262
Epoch 29/150
7200/7200 [==============================] - 0s 50us/step - loss: 0.4284 - acc: 0.8283
Epoch 30/150
7200/7200 [==============================] - 1s 93us/step - loss: 0.4293 - acc: 0.8278
Epoch 31/150
7200/7200 [==============================] - 1s 85us/step - loss: 0.4255 - acc: 0.8285
Epoch 32/150
7200/7200 [==============================] - 1s 87us/step - loss: 0.4261 - acc: 0.8274
Epoch 33/150
7200/7200 [==============================] - 1s 86us/step - loss: 0.4271 - acc: 0.8272
Epoch 34/150
7200/7200 [==============================] - 1s 86us/step - loss: 0.4250 - acc: 0.8281


7200/7200 [==============================] - 0s 37us/step - loss: 0.4242 - acc: 0.8044
Epoch 38/150
7200/7200 [==============================] - 0s 38us/step - loss: 0.4297 - acc: 0.8185
Epoch 39/150
7200/7200 [==============================] - 0s 50us/step - loss: 0.4261 - acc: 0.8226
Epoch 40/150
7200/7200 [==============================] - 1s 84us/step - loss: 0.4251 - acc: 0.8235
Epoch 41/150
7200/7200 [==============================] - 1s 82us/step - loss: 0.4281 - acc: 0.8235
Epoch 42/150
7200/7200 [==============================] - 1s 80us/step - loss: 0.4250 - acc: 0.8250
Epoch 43/150
7200/7200 [==============================] - 1s 80us/step - loss: 0.4252 - acc: 0.8276
Epoch 44/150
7200/7200 [==============================] - 1s 81us/step - loss: 0.4293 - acc: 0.8257
Epoch 45/150
7200/7200 [==============================] - 1s 82us/step - loss: 0.4256 - acc: 0.8244
Epoch 46/150
7200/7200 [==============================] - 1s 83us/step - loss: 0.4287 - acc: 0.8253
Epoch 47/150


7200/7200 [==============================] - 1s 85us/step - loss: 0.3868 - acc: 0.8332
Epoch 50/150
7200/7200 [==============================] - 1s 81us/step - loss: 0.3882 - acc: 0.8328
Epoch 51/150
7200/7200 [==============================] - 1s 83us/step - loss: 0.3891 - acc: 0.8349
Epoch 52/150
7200/7200 [==============================] - 1s 81us/step - loss: 0.3839 - acc: 0.8369
Epoch 53/150
7200/7200 [==============================] - 1s 82us/step - loss: 0.3834 - acc: 0.8367
Epoch 54/150
7200/7200 [==============================] - 1s 81us/step - loss: 0.3859 - acc: 0.8374
Epoch 55/150
7200/7200 [==============================] - 1s 82us/step - loss: 0.3884 - acc: 0.8349
Epoch 56/150
7200/7200 [==============================] - 0s 65us/step - loss: 0.3865 - acc: 0.8385
Epoch 57/150
7200/7200 [==============================] - 0s 37us/step - loss: 0.3816 - acc: 0.8401
Epoch 58/150
7200/7200 [==============================] - 0s 37us/step - loss: 0.3890 - acc: 0.8342
Epoch 59/150


7200/7200 [==============================] - 1s 102us/step - loss: 0.4205 - acc: 0.8301
Epoch 62/150
7200/7200 [==============================] - 0s 41us/step - loss: 0.4230 - acc: 0.8271: 0s - loss: 0.4104 - acc: 0
Epoch 63/150
7200/7200 [==============================] - 0s 38us/step - loss: 0.4185 - acc: 0.8317
Epoch 64/150
7200/7200 [==============================] - 0s 51us/step - loss: 0.4218 - acc: 0.8296
Epoch 65/150
7200/7200 [==============================] - 0s 39us/step - loss: 0.4167 - acc: 0.8312
Epoch 66/150
7200/7200 [==============================] - 0s 37us/step - loss: 0.4182 - acc: 0.8300
Epoch 67/150
7200/7200 [==============================] - 0s 54us/step - loss: 0.4197 - acc: 0.8317
Epoch 68/150
7200/7200 [==============================] - 0s 41us/step - loss: 0.4210 - acc: 0.8300
Epoch 69/150
7200/7200 [==============================] - 1s 81us/step - loss: 0.4193 - acc: 0.8286
Epoch 70/150
7200/7200 [==============================] - 1s 96us/step - loss: 0.420

7200/7200 [==============================] - 1s 84us/step - loss: 0.4245 - acc: 0.8339
Epoch 74/150
7200/7200 [==============================] - 1s 81us/step - loss: 0.4220 - acc: 0.8315
Epoch 75/150
7200/7200 [==============================] - 1s 81us/step - loss: 0.4192 - acc: 0.8321
Epoch 76/150
7200/7200 [==============================] - 1s 83us/step - loss: 0.4239 - acc: 0.8311
Epoch 77/150
7200/7200 [==============================] - 1s 71us/step - loss: 0.4194 - acc: 0.8276
Epoch 78/150
7200/7200 [==============================] - 0s 38us/step - loss: 0.4209 - acc: 0.8336
Epoch 79/150
7200/7200 [==============================] - 0s 37us/step - loss: 0.4228 - acc: 0.8306
Epoch 80/150
7200/7200 [==============================] - 0s 37us/step - loss: 0.4234 - acc: 0.8328
Epoch 81/150
7200/7200 [==============================] - 0s 37us/step - loss: 0.4229 - acc: 0.8319
Epoch 82/150
7200/7200 [==============================] - 0s 42us/step - loss: 0.4217 - acc: 0.8326
Epoch 83/150


Epoch 4/150
7200/7200 [==============================] - 0s 38us/step - loss: 0.4362 - acc: 0.7978
Epoch 5/150
7200/7200 [==============================] - 0s 39us/step - loss: 0.4367 - acc: 0.7978
Epoch 6/150
7200/7200 [==============================] - 0s 37us/step - loss: 0.4366 - acc: 0.7978
Epoch 7/150
7200/7200 [==============================] - 0s 37us/step - loss: 0.4357 - acc: 0.7978
Epoch 8/150
7200/7200 [==============================] - 0s 37us/step - loss: 0.4326 - acc: 0.7978
Epoch 9/150
7200/7200 [==============================] - 0s 37us/step - loss: 0.4321 - acc: 0.7978
Epoch 10/150
7200/7200 [==============================] - 0s 49us/step - loss: 0.4314 - acc: 0.7978
Epoch 11/150
7200/7200 [==============================] - 1s 85us/step - loss: 0.4288 - acc: 0.7978
Epoch 12/150
7200/7200 [==============================] - 1s 83us/step - loss: 0.4285 - acc: 0.7978
Epoch 13/150
7200/7200 [==============================] - 1s 83us/step - loss: 0.4280 - acc: 0.7978
Epoch 

7200/7200 [==============================] - 1s 84us/step - loss: 0.4305 - acc: 0.7993
Epoch 17/150
7200/7200 [==============================] - 1s 83us/step - loss: 0.4290 - acc: 0.7993
Epoch 18/150
7200/7200 [==============================] - 1s 82us/step - loss: 0.4236 - acc: 0.7993
Epoch 19/150
7200/7200 [==============================] - 0s 54us/step - loss: 0.4280 - acc: 0.7993
Epoch 20/150
7200/7200 [==============================] - 0s 40us/step - loss: 0.4295 - acc: 0.7993
Epoch 21/150
7200/7200 [==============================] - 0s 38us/step - loss: 0.4279 - acc: 0.7993
Epoch 22/150
7200/7200 [==============================] - 0s 43us/step - loss: 0.4255 - acc: 0.7993
Epoch 23/150
7200/7200 [==============================] - 0s 39us/step - loss: 0.4286 - acc: 0.8010
Epoch 24/150
7200/7200 [==============================] - 0s 38us/step - loss: 0.4285 - acc: 0.8190
Epoch 25/150
7200/7200 [==============================] - 0s 37us/step - loss: 0.4260 - acc: 0.8208
Epoch 26/150


7200/7200 [==============================] - 0s 40us/step - loss: 0.4184 - acc: 0.8340
Epoch 99/150
7200/7200 [==============================] - 0s 39us/step - loss: 0.4234 - acc: 0.8314
Epoch 100/150
7200/7200 [==============================] - 0s 38us/step - loss: 0.4196 - acc: 0.8325
Epoch 101/150
7200/7200 [==============================] - 0s 40us/step - loss: 0.4154 - acc: 0.8349
Epoch 102/150
7200/7200 [==============================] - 0s 39us/step - loss: 0.4191 - acc: 0.8336
Epoch 103/150
7200/7200 [==============================] - 0s 39us/step - loss: 0.4234 - acc: 0.8317
Epoch 104/150
7200/7200 [==============================] - 0s 40us/step - loss: 0.4197 - acc: 0.8324
Epoch 105/150
7200/7200 [==============================] - 1s 74us/step - loss: 0.4197 - acc: 0.8319
Epoch 106/150
7200/7200 [==============================] - 1s 86us/step - loss: 0.4233 - acc: 0.8314
Epoch 107/150
7200/7200 [==============================] - 1s 86us/step - loss: 0.4214 - acc: 0.8339
Epoch

7200/7200 [==============================] - 1s 88us/step - loss: 0.4219 - acc: 0.8287
Epoch 30/150
7200/7200 [==============================] - 1s 87us/step - loss: 0.4251 - acc: 0.8272
Epoch 31/150
7200/7200 [==============================] - 1s 85us/step - loss: 0.4229 - acc: 0.8289
Epoch 32/150
7200/7200 [==============================] - 1s 85us/step - loss: 0.4182 - acc: 0.8311
Epoch 33/150
7200/7200 [==============================] - 0s 54us/step - loss: 0.4234 - acc: 0.8287
Epoch 34/150
7200/7200 [==============================] - 0s 39us/step - loss: 0.4243 - acc: 0.8287
Epoch 35/150
7200/7200 [==============================] - 0s 39us/step - loss: 0.4198 - acc: 0.8300
Epoch 36/150
7200/7200 [==============================] - 0s 40us/step - loss: 0.4244 - acc: 0.8304
Epoch 37/150
7200/7200 [==============================] - 0s 39us/step - loss: 0.4239 - acc: 0.8300
Epoch 38/150
7200/7200 [==============================] - 0s 39us/step - loss: 0.4248 - acc: 0.8303
Epoch 39/150


Epoch 111/150
7200/7200 [==============================] - 0s 39us/step - loss: 0.4177 - acc: 0.8354
Epoch 112/150
7200/7200 [==============================] - 0s 40us/step - loss: 0.4217 - acc: 0.8319
Epoch 113/150
7200/7200 [==============================] - 0s 39us/step - loss: 0.4212 - acc: 0.8321
Epoch 114/150
7200/7200 [==============================] - 0s 64us/step - loss: 0.4202 - acc: 0.8315
Epoch 115/150
7200/7200 [==============================] - 1s 85us/step - loss: 0.4210 - acc: 0.8354
Epoch 116/150
7200/7200 [==============================] - 1s 93us/step - loss: 0.4220 - acc: 0.8329
Epoch 117/150
7200/7200 [==============================] - 1s 85us/step - loss: 0.4217 - acc: 0.8331
Epoch 118/150
7200/7200 [==============================] - 1s 113us/step - loss: 0.4228 - acc: 0.8347
Epoch 119/150
7200/7200 [==============================] - 1s 110us/step - loss: 0.4212 - acc: 0.8365
Epoch 120/150
7200/7200 [==============================] - 1s 102us/step - loss: 0.4204 -

In [88]:
accuracy.mean()

0.8356249935925006

In [89]:
accuracy.std()

0.014353245739509529

The mean accuracy with a neural network model is around 83.5% and also the mean variance is quite low. Thus, the model generalized well on the testing data.

### Classification using a Decision Tree 

In [31]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score
dtree = DecisionTreeClassifier(max_depth= 8)

In [69]:
d_class = dtree.fit(x_train, y_train)

In [70]:
y_dpred = d_class.predict(x_test)

In [71]:
cm2 = confusion_matrix(y_test, y_dpred)

In [72]:
cm2

array([[1522,   65],
       [ 220,  193]])

#### Applying Cross validation

In [77]:
acc = cross_val_score(estimator=d_class, X = x_train, y= y_train, cv=10, n_jobs=-1)

In [80]:
acc.mean()

0.8528782539113344

In [81]:
acc.std()

0.006912189924639332

The mean accuracy with a Decision Tree model is around 85.3% and also the mean variance is lower than the previous one. Thus, the model generalized well on the testing data even better.

### Parameter Tuning

In [33]:
from sklearn.model_selection import GridSearchCV

In [35]:
def build_classifier(optim):
    classifier = Sequential()
    # input Layer and first inner layer with 6 nodes
    classifier.add(Dense(input_dim = 11, units = 6, activation = 'relu', kernel_initializer = 'uniform' ))
    classifier.add(Dropout(rate = 0.1))
    # second Layer 
    classifier.add(Dense(units = 8, activation = 'relu', kernel_initializer = 'uniform' ))
    classifier.add(Dropout(rate = 0.2))
    # output Layer
    classifier.add(Dense(units = 1, activation = 'sigmoid', kernel_initializer = 'uniform' ))

    classifier.compile(optimizer = optim, loss = 'binary_crossentropy', metrics = ['accuracy'])
    return classifier

classifier = KerasClassifier(build_fn=build_classifier, shuffle = True)

In [36]:
parameters = {'epochs' : [100, 150, 500],
             'batch_size' : [25, 32, 64],
             'optim' : ['adam', 'rmsprop']}

In [37]:
gridsearch = GridSearchCV(estimator=classifier, param_grid=parameters, scoring= 'accuracy', cv = 10)

In [ ]:
gridsearch = gridsearch.fit(X = x_train, y = y_train )

In [ ]:
best_params = gridsearch.best_params_
best_acc = gridsearch.best_score_